### PR TITLE
Add readme and swagger specification

### PR DESCRIPTION
**Pull request by ianbrown9475**

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -1,75 +1,298 @@
...
+      summary: "Retrieve cards based on search parameters"
+      description: ""
+      operationId: "getCards"
+      consumes:
+      - "application/json"
+      produces:
+      - "application/json"
       parameters:
-        - $ref: '#/parameters/pretty'
+        - name: "id"
```
Comment (jaredkosanovic) [inline swagger.yaml]: Why have an `id` query parameter if you also have a /cards/`id` endpoint? You don't have to change anything, just curious

File: swagger.yaml [hunk 2]
```diff
@@ -1,75 +1,298 @@
...
         200:
-          description: Object containing information
+          description: "Cards successfully retrieved"
+          schema:
+            $ref: "#/definitions/MultiData"
+        400:
+          description: "Invalid request"
+          schema:
+            $ref: "#/definitions/Error"
+        404:
```
Comment (jaredkosanovic) [inline swagger.yaml]: I'm not sure if this endpoint should return a 404. In jsonAPI spec, endpoints that usually return an array of objects would return a 200 with an empty `"data"` array if no cards were found that matched the query. http://jsonapi.org/format/#fetching-resources-responses-200

File: swagger.yaml [hunk 3]
```diff
@@ -1,75 +1,298 @@
...
-  /:
+  /cards:
+
     get:
-      summary: Get information
-      description: Build and runtime information about the application.
+      tags:
+      - "cards"
+      summary: "Retrieve cards based on search parameters"
+      description: ""
```
Comment (jaredkosanovic) [inline swagger.yaml]: Looks like you aren't using the description fields for any of the endpoints. Could you remove them instead?

File: swagger.yaml [hunk 4]
```diff
@@ -1,75 +1,298 @@
...
+      parameters:
+        - name: "body"
+          in: "body"
+          description: "Card object to be added to the database"
+          required: true
+          schema:
+            $ref: "#/definitions/SingleData"
+
+      responses:
+        201:
```
Comment (jaredkosanovic) [inline swagger.yaml]: For POST and PUT, it would be a good idea to return the object that was added or updated in the http response.

File: swagger.yaml [hunk 5]
```diff
@@ -1,75 +1,298 @@
...
+          format: "int64"
+
+        - name: "body"
+          in: "body"
+          description: "Object that will overwrite the card specified by its id"
+          required: true
+          schema:
+            $ref: "#/definitions/SingleData"
+      responses:
+        204:
```
Comment (jaredkosanovic) [inline swagger.yaml]: Same comment as the POST, it would be good to return the updated object in the response. I also made a mistake with my comment in Jira when recommending you change the status code. I realized a 200 might be the better option because there isn't a good code for "updated".

File: swagger.yaml [hunk 6]
```diff
@@ -1,75 +1,298 @@
...
+        - name: "id"
+          in: "path"
+          description: "Id of card to be removed"
+          required: true
+          type: "integer"
+          format: "int64"
+      responses:
+        204:
+          description: "Card successfully removed"
+        400:
```
Comment (jaredkosanovic) [inline swagger.yaml]: What type of request might return a `400` for a `DELETE`?

File: swagger.yaml [hunk 7]
```diff
@@ -1,75 +1,298 @@
...
+
+        - name: "energyMax"
+          in: "query"
+          description: "Maximum value of energy costs of cards (X = 0). Used with energyMin"
+          required: false
+          type: "integer"
+
+        - name: "keywords"
+          in: "query"
+          description: "List of keywords found in a card's description"
```
Comment (jaredkosanovic) [inline swagger.yaml]: Just curious, what will this query parameter be searching? I didn't see a description field for the card object.

**Review by jaredkosanovic — CHANGES_REQUESTED** [review]

**Comment by ianbrown9475** [comment]
- id parameter removed (redundant due to /cards/{id} endpoint

**Comment by ianbrown9475** [comment]
- Removed id parameter in GET /cards due to redundancy with /cards/{id} endpoint
- GET /cards no longer returns a 404 due to no cards matching
- Removed empty description fields
- PUT, POST, and DELETE now respond with an HTTP 200 and return the SingleData object that was updated, added, or removed respectively
- DELETE no longer returns an HTTP 400 because there is no syntax in the body that can be incorrect
- Added a "description" field in the Card object
- Added an isRandom parameter (boolean) in the GET /cards endpoint that is meant to return cards in a random order if true. This is meant to emulate drawing cards of a certain type, color, ...
- Added NewData object used by the POST /cards endpoint that lacks an id field for adding a new card. Structure is otherwise equivalent to SingleData
- Added an HTTP 500 response to all endpoints in the event of an internal server error

File: swagger.yaml [hunk 8]
```diff
@@ -1,75 +1,330 @@
...
+      produces:
+      - "application/json"
       parameters:
-        - $ref: '#/parameters/pretty'
+        - name: "body"
+          in: "body"
+          description: "Card object to be added to the database"
+          required: true
+          schema:
+            $ref: "#/definitions/SingleData"
```
Comment (jaredkosanovic) [inline swagger.yaml]: Should this be using the `NewData` object too, similar to put?

File: swagger.yaml [hunk 9]
```diff
@@ -1,75 +1,330 @@
...
+
       responses:
         200:
```
Comment (jaredkosanovic) [inline swagger.yaml]: Perhaps a `201` would be more appropriate when creating a new card object? Sorry if my comment didn't make sense originally.

File: swagger.yaml [hunk 10]
```diff
@@ -1,75 +1,330 @@
...
+
+  NewData:
+    type: "object"
+    properties:
+      data:
+        type: "object"
+        properties:
+          type:
+            type: "string"
+          attributes:
```
Comment (jaredkosanovic) [inline swagger.yaml:314]: The attributes for NewData and Card are the same. Maybe use one shared object instead?

File: swagger.yaml [hunk 11]
```diff
@@ -240,14 +244,16 @@ paths:
           type: "integer"
           format: "int64"
       responses:
-        204:
+        200:
```
Comment (jaredkosanovic) [inline swagger.yaml]: Just curious, why did you change this from a 204? I would say a 204 is more common when deleting a resource, but either are acceptable.

**Review by jaredkosanovic — CHANGES_REQUESTED** [review]

**Comment by ianbrown9475** [comment]
- POST /cards now uses NewData in body instead of PUT /cards/{id}
- PUT /cards/{id} uses SingleData in compliance with jsonapi: http://jsonapi.org/format/#crud-updating
- POST /cards responds with a 201 rather than a 200 for resource creation
- Restructured objects to factor properties better
- Successful DELETE responds with a 204 and no content rather than responding with a 200 and including the deleted object

**Review by jaredkosanovic — APPROVED** [review]
